docker-compose.yml

```diff
@@ -1,17 +1,19 @@
-web:
-  image: colinxfleming/dcaf_case_management
-  command: rails s -p 3000 -b 0.0.0.0
-  volumes:
-    - .:/usr/src/app
-  ports:
-    - "3000:3000"
-  links:
-    - db
-  environment:
-    mongohost: db
-  env_file:
-    - web-variables.env
-db:
-  image: mongo
-  ports:
-    - "27017:27017"
+version: '3.1'
+services:
+  web:
+    build: .
```

Comment:
What does having both the `image` and `build` do? Normally `docker-compose build` refs the `:latest` image and then `docker-compose up` pulls the layers. (My own finger muscle memory when using docker tends to be `git pull origin master && docker-compose build && docker-compose up` or some form of that.) It looks like it falls back on a `build`, but wanted to make sure that's what's supposed to happen.

Reply:
In the 3.1 version, if you specify build and image it:

* Tells the `docker-compose build` command what image name to use when building the image, translates to basically `docker build -t colinxfleming/dcaf_case_management:latest .`

The main reason I did that was to just ensure that if someone builds it knows to a: create the image under the same namespace (avoid a bunch of odd images/layers being created) b: that it for sure would build, and not just always pull the remote from Dockerhub

Reply:
Trying to follow this convo...

Are you saying that `docker-compose` uses the `image` parameter as a name (and nothing else) when you have both `image` and `build` params?

B/c it seems really weird if...

*Only `image` tag*: Build from remote, using `colinxfleming/dcaf_case_management:latest` ans a reference.
*`image` and `build` tag*: Build from local, using `colinxfleming/dcaf_case_management:latest`as a name.

Reply:
@colinxfleming
[Docker docs (ha, Docker docs) say this about `dockercompose.yml` files](https://docs.docker.com/compose/compose-file/#image):

> 
> image
> 
> Specify the image to start the container from. Can either be a repository/tag or a partial image ID.
> ```
> image: redis
> image: ubuntu:14.04
> image: tutum/influxdb
> image: example-registry.com:4000/postgresql
> image: a4bc65fd
> ```
> If the image does not exist, Compose attempts to pull it, unless you have also specified build, in which case it builds it using the specified options and tags it with the specified tag.

So as I understand it,
`docker-compose pull` will remotely pull any service that has an `image` property.
`docker-compose build` will locally build any service that has a `build` property.
`docker-compose up` will attempt to use existing images. If images do not exist, then it will build images when a `build` property is specified. And it will pull images when *only* a `pull` property is specified.

And as @DarthHater mentioned, if `build` and `image` property are specified together, then the `build` property uses `image` tag for naming purposes. It doesn't seem like there's any fallback behavior anywhere...

Reply:
As with all things, it's always more complex than a simple example BUT:

* `docker-compose up` will only build an image if it has to, if existing images are there that satisfy it's requirements, it'll just use them or the containers it's previously created from them
* `docker-compose pull` will pull from the Docker hub, remote, so if for some reason you want whatever is out there it's entirely possible to get it
* In most cases you don't absolutely need what's out on the hub BECAUSE:
  * The biggest advantage the remote image has it's got Gems already installed. I wouldn't say 9 times out of 10 but fairly often you are either introducing a new Gem or something akin and likely need to do `bundle install` to get things up and running

Build first would actually be better in most cases, because of the aforementioned Gem updates. That'll happen as you are building. However say you have an image locally already (lots of people do from pulling it in the past, and would in the future), the entrypoint.sh addition just makes it sort of seemless in terms of making sure your Gems are all there.

Reply:
It's news to me that you can have both `image` and `build` commands in a dockerfile and that it won't throw it back in your face! A nice-to-have feature for sure. I'm good with this part. Note to self to start internalizing `docker-compose pull` I guess.